source/containerimage/pull.go

```diff
@@ -222,22 +222,55 @@ func (p *puller) Snapshot(ctx context.Context) (cache.ImmutableRef, error) {
 	}
 	stopProgress()
 
+	var usedBlobs, unusedBlobs []ocispec.Descriptor
+
 	if schema1Converter != nil {
+		ongoing.remove(p.desc) // Not left in the content store so this is sufficient.
 		p.desc, err = schema1Converter.Convert(ctx)
 		if err != nil {
 			return nil, err
 		}
+		ongoing.add(p.desc)
+
+		var mu sync.Mutex // images.Dispatch calls handlers in parallel
+		allBlobs := make(map[digest.Digest]ocispec.Descriptor)
+		for _, j := range ongoing.added {
+			allBlobs[j.Digest] = j.Descriptor
+		}
+
+		handlers := []images.Handler{
+			images.HandlerFunc(func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+				usedBlobs = append(usedBlobs, desc)
+				mu.Lock()
```

Comment:
I think this is the source of the race, the `append` should be under the lock too since these handlers are run in parallel and without the lock we risk missing entries out of `usedBlobs` due to a racing `append`s.

Will run a few more iterations of the test case to check if it has really been fixed and then push the fix.